builder/dockerfile/dispatchers.go

```diff
@@ -210,20 +210,44 @@ func (d *dispatchRequest) getExpandedImageName(shlex *shell.Lex, name string) (s
 	}
 	return name, nil
 }
-func (d *dispatchRequest) getImageOrStage(name string) (builder.Image, error) {
+
+// getOsFromFlagsAndStage calculates the operating system if we need to pull an image.
+// stagePlatform contains the value supplied by optional `--platform=` on
+// a current FROM statement. b.builder.options.Platform contains the operating
+// system part of the optional flag passed in the API call (or CLI flag
+// through `docker build --platform=...`).
+func (d *dispatchRequest) getOsFromFlagsAndStage(stagePlatform string) string {
+	osForPull := ""
+	// First, take the API platform if nothing provided on FROM
+	if stagePlatform == "" && d.builder.options.Platform != "" {
+		osForPull = d.builder.options.Platform
+	}
+	// Next, use the FROM flag if that was provided
+	if osForPull == "" && stagePlatform != "" {
+		osForPull = stagePlatform
+	}
+	// Finally, assume the host OS
+	if osForPull == "" {
+		osForPull = runtime.GOOS
+	}
+	return osForPull
```

Comment:
This is difficult to follow. I think this would be more clear:

```go
switch {
case stagePlatform != "":
    return stagePlatform
case d.builder.options.Platform != "":
    return d.builder.options.Platform
default:
    return runtime.GOOS
}
```

Reply:
OK, can simplify, but your logic above isn't quite correct. Priority is the d.builder.options.Platform

Reply:
I agree the `options.Platform` should be the highest priority, so I'm glad we're making that change.

However that's not what your code does, or what it says in the document:

> If both the API and a `FROM` statement both contain a platform, the one in the FROM statement will be used.

Reply:
Actually, not thinking this morning. You were right the first time. Will push an update shortly